counterparty-core/counterpartycore/lib/messages/move.py

```diff
@@ -0,0 +1,84 @@
+import logging
+
+from counterpartycore.lib import config, exceptions, ledger, script, util
+
+logger = logging.getLogger(config.LOGGER_NAME)
+
+
+def compose(db, source, destination):
+    # We don't check if the source has attached assets
+    # to allow transaction chaining in the same block
+
+    if not util.is_utxo_format(source):
+        raise exceptions.ComposeError("Invalid source utxo format")
+
+    try:
+        script.validate(destination)
+    except script.AddressError as e:
+        raise exceptions.ComposeError("destination must be an address") from e
+
+    return (source, [(destination, None)], None)
+
+
+# call on each transaction
+def move_assets(db, tx):
+    if "utxos_info" not in tx or not tx["utxos_info"]:
+        return
+
+    sources, destination, _outputs_count, _op_return_output = util.parse_utxos_info(
+        tx["utxos_info"]
+    )
+
+    # do nothing if no source or destination
+    if len(sources) == 0 or not destination:
```

Comment:
If no valid destination in a move transaction, either burning all the assets or returning them to the address associated with the UTXO to avoid assets remaining in a spent UTXO

` if len(sources) == 0 or not destination:
        return`

Reply:
Why does it bother you so much to have assets attached to an unspendable output? It's actually a kind of burning! I mean if someone goes to the effort of crafting a transaction with a single OP_RETURN it's because they want it to become impossible to move assets and that's the definition of burning.

Reply:
My main considerations are:

1. Assets bound to spent UTXOs are essentially garbage data and unnecessary, as users cannot actually use them, which may also cause some confusion.
2. Third-party platforms (such as OpenStamp) need to determine whether assets are bound to spent UTXOs when calculating users’ valid assets; we don’t want to display unusable assets to users.


Of course, these are just my suggestions for everyone’s discussion and reference.

Reply:
Yeah, I've thought a lot about this and I do think that preventing burning of UTXO–attached assets is probably a good idea. There are other ways to burn.